### ISSUE TITLE
Roundup: [oss-security] libpcre: invalid memory read in match (pcre_exec.c)

### ISSUE THREAD

**Issue by grahamc** [NixOS/security#57]
Here is a report from the oss-security mailing list for [Vulnerability Roundup 27](https://github.com/NixOS/nixpkgs/issues/24319).

[Skip to First Email](#first-email)

# Instructions:
## Identification
Identify if we have the software, in 16.09, 17.03, and unstable. Then determine if we are vulnerable, and make a comment with your findings. It can also be helpful to specify if you think there is a patch, or if it can be fixed via a general update.

Example:

```
unstable: we are not vulnerable (link to the package)
17.03: we are vulnerable (link to the package)
16.09: we don't have it packaged
```

**IMPORTANT**: If you believe there are possibly related issues, bring them up on the _parent_ issue!

## Patching
Start by commenting on this issue saying you're working on a patch. This way, we don't duplicate work.

If you open a pull request, tag this issue _and_ the master issue for the roundup.

If you commit the patch directly to a branch, please leave a comment on this issue with the branch and the commit hash, example:

```
fixed:

release-16.09: abc123
```

[Skip to First Email](#first-email)

## Upon Completion ...
 - [ ] Update Graham's database

## Info

_Triage Indicator:_

```
-needs-triage +roundup27 thread:000000000000425d
```

 - File Search: https://search.nix.gsc.io/?q=match&i=fosho&repos=nixos-nixpkgs
 - GitHub Search: https://github.com/NixOS/nixpkgs/search?utf8=%E2%9C%93&q=match+in%3Apath&type=Code

Should the search term be changed from `match`? Suggest a new package search by commenting:

```
-suggested:match +suggested:correctPackageName thread:000000000000425d
```

Known CVEs: CVE-2017-7186

---

<a name="first-email"></a>
[Skip to End](#end)

##### Mon, 20 Mar 2017 10:26:26 +0000 "Agostino Sarubbo" <ago-at-gentoo.org>, `762448.990109505-sendEmail@localhost`

    Description:
    libpcre is a perl-compatible regular expression library.
    
    A fuzz on libpcre1 through the pcretest utility revealed an invalid read in the library. For who is interested in a detailed description of the bug, will follow a feedback from upstream:
    
    This was a genuine bug in the 32-bit library. Thanks for finding it. The crash was caused by trying to find a Unicode property for a code value greater than 0x10ffff, the Unicode maximum, when running 
    in non-UTF mode (where character values can be up to 0xffffffff). The bug was in both PCRE1 and PCRE2. I have fixed both of them.
    
    The complete ASan output:
    
    # pcretest -32 -d $FILE
    ==14788==ERROR: AddressSanitizer: SEGV on unknown address 0x7f1bbffed4df (pc 0x7f1bbee3fe6b bp 0x7fff8b50d8c0 sp 0x7fff8b50d3a0 T0)
    ==14788==The signal is caused by a READ memory access.
        #0 0x7f1bbee3fe6a in match /tmp/portage/dev-libs/libpcre-8.40/work/pcre-8.40/pcre_exec.c:5473:18
        #1 0x7f1bbee09226 in pcre32_exec /tmp/portage/dev-libs/libpcre-8.40/work/pcre-8.40/pcre_exec.c:6936:8
        #2 0x527d6c in main /tmp/portage/dev-libs/libpcre-8.40/work/pcre-8.40/pcretest.c:5218:9
        #3 0x7f1bbddd678f in __libc_start_main /tmp/portage/sys-libs/glibc-2.23-r3/work/glibc-2.23/csu/../csu/libc-start.c:289
        #4 0x41b438 in _init (/usr/bin/pcretest+0x41b438)
    
    AddressSanitizer can not provide additional info.
    SUMMARY: AddressSanitizer: SEGV /tmp/portage/dev-libs/libpcre-8.40/work/pcre-8.40/pcre_exec.c:5473:18 in match
    ==14788==ABORTING
    
    Affected version:
    8.40 and 10.23
    
    Fixed version:
    8.41 and 10.24 (not released atm)
    
    Commit fix for libpcre1:
    https://vcs.pcre.org/pcre/code/trunk/pcre_internal.h?r1=1649&r2=1688&sortby=date
    https://vcs.pcre.org/pcre/code/trunk/pcre_ucd.c?r1=1490&r2=1688&sortby=date
    
    Commit fix for libpcre2:
    https://vcs.pcre.org/pcre2/code/trunk/src/pcre2_ucd.c?r1=316&r2=670&sortby=date
    https://vcs.pcre.org/pcre2/code/trunk/src/pcre2_internal.h?r1=600&r2=670&sortby=date
    
    Credit:
    This bug was discovered by Agostino Sarubbo of Gentoo.
    
    CVE:
    CVE-2017-7186
    
    Reproducer:
    https://github.com/asarubbo/poc/blob/master/00204-pcre-invalidread1-pcre_exec
    
    Timeline:
    2017-02-23: bug discovered and reported to upstream
    2017-02-24: upstream released a patch
    2017-03-14: blog post about the issue
    2017-03-19: CVE assigned
    
    Note:
    This bug was found with American Fuzzy Lop.
    
    Permalink:
    https://blogs.gentoo.org/ago/2017/03/14/libpcre-invalid-memory-read-in-match-pcre_exec-c
    
    --
    Agostino Sarubbo
    Gentoo Linux Developer
    
    

[Skip to End](#end)

---

<a name="end"></a>

**Comment by ndowens** [comment]
master: 8.39
16.09: 8.39
17.03: 8.39

Upstream has not released a version to fix this yet

**Comment by globin** [comment]
Working on it

**Comment by globin** [comment]
pcre: fix in staging, 16.09, 17.03
pcre2: fix on master, 16.09, 17.03

**Comment by FRidh** [comment]
Note that the fix on 16.09 caused about 5000 failures
http://hydra.nixos.org/eval/1344772?full=1#tabs-inputs

**Comment by globin** [comment]
> I'm pretty sure that mostly only ghc timed out but the log isn't available anymore so I can't check but I'll build that in a sec and then restart the builds.
